File: dndapi-onion/Domain/Domain/PlayArea/TokenFactory.cs
using Domain.Dto.Shared;

namespace Domain.Domain.PlayArea
{
    public static class TokenFactory
    {
        public static Token Create(TokenDto dto)
        {
            switch (dto.Type)
            {
                case TokenType.Character:
                    return new CharacterToken(dto as CharacterTokenDto);
                default:
                    return null;
            }
        }
    }
}

File: angular/src/app/shared/models/play-area/token-factory.ts
import { JournalCharacterSheet } from '../journal/journalitems/journal-character-sheet.model';
import { CharacterToken } from './character-token.model';
import { DefaultTokenDto } from './backend/default-token.dto';
import { DefaultToken } from './default-token.model';
import { CharacterTokenDto } from './backend/character-token.dto';

export abstract class TokenFactory {
    static create(x: number, y: number, journalItem: JournalCharacterSheet) {
        const dto = new CharacterTokenDto();
        dto.characterSheetId = journalItem.id;
        dto.x = x;
        dto.y = y;
        const token = new CharacterToken(dto as CharacterTokenDto);
        return token;
    }

    static createFromDto(dto: DefaultTokenDto): DefaultToken {
        const token = new CharacterToken(dto as CharacterTokenDto);
        return token;
    }
}
